### ISSUE TITLE
ServiceInstance.servicecatalog.k8s.io \"\" is invalid: metadata.name: Required value: name or generateName is required

### ISSUE THREAD

**Issue by cmoulliard** [snowdrop/component-operator#47]
## Issue

When we deploy the following resources on k8s

```
---
apiVersion: "v1"
kind: "List"
items:
- apiVersion: "devexp.runtime.redhat.com/v1alpha2"
  kind: "Capability"
  metadata:
    name: "postgresql-db"
  spec:
    class: "dh-postgresql-apb"
    plan: "dev"
    secretName: "postgresql-db"
    parameters:
    - name: "postgresql_user"
      value: "luke"
    - name: "postgresql_password"
      value: "secret"
    - name: "postgresql_database"
      value: "my_data"
    - name: "postgresql_version"
      value: "9.6"
- apiVersion: "devexp.runtime.redhat.com/v1alpha2"
  kind: "Component"
  metadata:
    labels:
      app: "fruit-backend-sb"
      version: "0.0.1-SNAPSHOT"
      group: "dabou"
    name: "fruit-backend-sb"
  spec:
    deploymentMode: "dev"
    runtime: "spring-boot"
    version: "2.1.3.RELEASE"
    exposeService: true
    envs:
    - name: "SPRING_PROFILES_ACTIVE"
      value: "openshift-catalog"
- apiVersion: "devexp.runtime.redhat.com/v1alpha2"
  kind: "Link"
  metadata:
    name: "link-to-database"
  spec:
    name: "link-to-database"
    componentName: "fruit-client-sb"
    kind: "Secret"
    ref: "postgresql-db"
```

then we got a `ServiceInstance Error` as you can see hereafter

```
{"level":"error","ts":1559746096.4847112,"logger":"capability.controller","msg":"Failed to create new ","Namespace":"demo","kind":"ServiceInstance","Namespace":"demo","error":"ServiceInstance.servicecatalog.k8s.io \"\" is invalid: metadata.name: Required value: name or generateName is required","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error
\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/snowdrop/component-operator/pkg/controller/capability.(*ReconcileCapability).create\n\t/go/src/github.com/snowdrop/component-operator/pkg/controller/capability/controller.go:114
github.com/snowdrop/component-operator/pkg/controller/capability.(*ReconcileCapability).Reconcile
\t/go/src/github.com/snowdrop/component-operator/pkg/controller/capability/controller.go:174
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.9/pkg/internal/controller/controller.go:215
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.9/pkg/internal/controller/controller.go:158
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func
\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181126123746-eddba98df674/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntil
\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181126123746-eddba98df674/pkg/util/wait/wait.go:134
k8s.io/apimachinery/pkg/util/wait.Until
\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181126123746-eddba98df674/pkg/util/wait/wait.go:88"}
{"level":"error","ts":1559746096.4848354,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"service-controller","request":"demo/postgresql-db","error":"ServiceInstance.servicecatalog.k8s.io \"\" is invalid: metadata.name: Required value: name or generateName is required","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error
\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.9/pkg/internal/controller/controller.go:217
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.9/pkg/internal/controller/controller.go:158
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func
\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181126123746-eddba98df674/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntil
\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181126123746-eddba98df674/pkg/util/wait/wait.go:13
k8s.io/apimachinery/pkg/util/wait.Until
\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181126123746-eddba98df674/pkg/util/wait/wait.go:88"}
```

This problem comes from the fact that no `Capability.spec.name` is defined by ap4k.
So the resource created should be

```
- apiVersion: "devexp.runtime.redhat.com/v1alpha2"
  kind: "Capability"
  metadata:
    name: "postgresql-db"
  spec:
      name: "postgresql-db"
      class: "dh-postgresql-apb"
      plan: "dev"
      secretName: "postgresql-db"
      parameters:
      - name: "postgresql_user"
        value: "luke"
      - name: "postgresql_password"
        value: "secret"
      - name: "postgresql_database"
        value: "my_data"
      - name: "postgresql_version"
        value: "9.6"
```

**REMARK**: We could decide together that if ap4k don't include the name under `.spec.name` we could use `.metadata.name`. To be discussed ? WDYT @metacosm @aureamunoz

**Comment by aureamunoz** [comment]
We have adopted the convention of using the `.metadata.name` as indicated in the previous remark (view [issue #48](https://github.com/snowdrop/component-operator/issues/48)) so this fix is to do on server side (component-operator) in order to get the information from  `.metadata.name`.

**Comment by cmoulliard** [comment]
Fixed with commit pushed for #48
